### PR TITLE
Migrating search experiments towards our new A/B Testing framework

### PR DESCRIPTION
**Pull request by leogrim**
@stkem @ymatsuda @yingjieMiao

### CODE REVIEW

**Comment by leogrim** [comment]
Note: at the first glance, this probably looks more convoluted that it used to. But going forward we think that separating the experiment assignment logic from the logic that modifies the actual behavior based on available experiments (in this case, looking up the right SearchConfig) should pay off.

File: kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala [hunk 1]
```diff
@@ -98,59 +107,47 @@ class SearchConfigManager(configDir: Option[File], shoeboxClient: ShoeboxService
...
-    (hash - min) / (max - min)
-  }
+  private var userConfig = Map.empty[Id[User], SearchConfig]
+  def getUserConfig(userId: Id[User]) = userConfig.getOrElse(userId, defaultConfig)
+  def setUserConfig(userId: Id[User], config: SearchConfig) { userConfig = userConfig + (userId -> config) }
+  def resetUserConfig(userId: Id[User]) { userConfig = userConfig - userId }
 
-
-  def getConfig(userId: Id[User], queryText: String, excludeFromExperiments: Boolean = false): (SearchConfig, Option[Id[SearchConfigExperiment]]) = {
+  def getConfig(userId: Id[User], userExperiments: Set[ExperimentType]): (SearchConfig, Option[Id[SearchConfigExperiment]]) = {
```
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala:136]: Looks like once we are here, the assignment is deterministic. The random factor has been moved to internByName, I assume? For each user, we persist his search experiment(s) in UserExperimentRepo? So, every time we roll out some search experiments, we will touch lots of records in UserExperimentRepo?

File: kifi-backend/modules/shoebox/app/com/keepit/model/ProbabilisticExperimentGeneratorRepo.scala [hunk 2]
```diff
@@ -49,5 +50,15 @@ class ProbabilisticExperimentGeneratorRepoImpl @Inject()(
     val q = (for(f <- rows if f.name === name && f.state =!= exclude.orNull) yield f)
     q.firstOption
   }
+
+  def internByName(name: Name[ProbabilisticExperimentGenerator], density: ProbabilityDensity[ExperimentType], salt: Option[String], condition: Option[ExperimentType])(implicit session: RWSession): ProbabilisticExperimentGenerator = {
+    val defaultSalt = name.name + "Salt"
```
Comment (yingjieMiao) [inline kifi-backend/modules/shoebox/app/com/keepit/model/ProbabilisticExperimentGeneratorRepo.scala:55]: Still not clear what salt stands for. Somehow related to "density"? A tag?

**Comment by yingjieMiao** [comment]
lgtm overall

File: kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala:136]: It's not UserExperimentRepo but ProbabilisticExperimentGeneratorRepo, it's basically a generic version of the sampling that was done for search experiments, available to all.

File: kifi-backend/modules/shoebox/app/com/keepit/model/ProbabilisticExperimentGeneratorRepo.scala (hunk 2 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/model/ProbabilisticExperimentGeneratorRepo.scala:55]: Look up ProbabilisticExperimentGenerator, the salt is mixed with userIds when "sampling", this way when we experiment with 10% of users, it's not always the same 10% who get the various experiments.

File: kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala (hunk 1 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala:136]: There is no central repo to _persist_ all experiments for a user?

If I have an authenticated request, then request.experiments would be userExperiments for that user + something generated from the randomGenerator, computed on-the-fly?

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala:136]: Exactly!

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala [hunk 3]
```diff
@@ -163,11 +163,11 @@ object Search extends Service {
     def friendMapJson(userId: Id[User], query: Option[String] = None, minKeeps: Option[Int] = None) = ServiceRoute(GET, "/internal/search/search/friendMapJson", Param("userId", userId), Param("query", query), Param("minKeeps", minKeeps))
     def search(
         userId: Id[User],
-        noSearchExperiments: Boolean,
+        experiments: Set[ExperimentType],
         acceptLangs: Seq[String],
         rawQuery: String) = {
         val params = "userId=" + userId.id.toString +
-                     "&nse=" + noSearchExperiments +
+                     "&exp=" + experiments.mkString(",") +
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala]: Why send all the experiments along to search? Could we just have search get them from the Commander when needed (where they might also be cached)?

File: kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala [hunk 4]
```diff
@@ -87,10 +88,18 @@ object SearchConfig {
...
 
+  val empty = new SearchConfig(Map.empty)
   val defaultConfig = new SearchConfig(SearchConfig.defaultParams)
 
   def apply(params: (String, String)*): SearchConfig = SearchConfig(Map(params:_*))
   def getDescription(name: String) = descriptions.get(name)
+
+  def byUserSegment(seg: UserSegment): SearchConfig = {
+    seg.value match {
+      case 3 => new SearchConfig(Map("dampingHalfDecayFriends" -> "2.5", "percentMatch" -> "85"))
```
Comment (stkem) [inline kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala:99]: Only `3` does something different?

Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala:99]: yes

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala]: This is an internal route (T-search I believe), so we would resolve the experiments twice?

File: kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala (hunk 4 above)
Comment (yingjieMiao) [inline kifi-backend/modules/common/app/com/keepit/search/SearchConfig.scala:99]: That is inherited from old code. I was asked to do it only for that group of user.

**Comment by stkem** [comment]
Made a few comments. lgtm otherwise.
We have to be careful with deploy order because of the changed search call.

File: kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala (hunk 3 above)
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/routes/Router.scala]: @stkem gone

**Pull request merged**